pkg/helm/watcher/cache/cache.go

```diff
@@ -0,0 +1,182 @@
+package cache
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/gofrs/flock"
+
+	pb "github.com/weaveworks/weave-gitops/pkg/api/profiles"
+)
+
+const (
+	// we globally lock the whole base cache folder to avoid the
+	// possibility of checking if a sub-folder exists, just so it is created the moment
+	// we try to create it after we checked that it exists.
+	lockFilename    = "cache.lock"
+	profileFilename = "profiles.yaml"
+	valuesFilename  = "values.yaml"
+)
+
+// type alias for easier reading of the cache layout for values.
+type profileName = string
+type profileVersion = string
+
+// ValueMap contains easy access for a profile name and version based values file.
+type ValueMap map[profileName]map[profileVersion][]byte
+
+// Cache defines an interface to work with the profile data cacher.
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+//counterfeiter:generate . Cache
+type Cache interface {
+	Update(helmRepoNamespace, helmRepoName string, value Data) error
+	// GetProfiles specifically retrieve profiles data only to avoid traversing the values structure for no reason.
+	GetProfiles(helmRepoNamespace, helmRepoName string) ([]*pb.Profile, error)
+	// GetProfileValues will try and find a specific values file for the given profileName and profileVersion. Returns an
+	// error if said version is not found.
+	GetProfileValues(helmRepoNamespace, helmRepoName, profileName, profileVersion string) ([]byte, error)
+}
+
+// Data is explicit data for a specific profile including values.
+// Saved as `profiles.yaml` and `profileName/version/values.yaml`.
+type Data struct {
+	Profiles []*pb.Profile `yaml:"profiles"`
+	Values   ValueMap
+}
+
+// HelmCache is used to cache profiles data from scanner helm repositories.
+type HelmCache struct {
+	cacheDir string
+}
+
+var _ Cache = &HelmCache{}
+
+// NewCache initialises the cache and returns it.
+func NewCache(cacheDir string) (*HelmCache, error) {
+	if _, err := os.Stat(cacheDir); os.IsNotExist(err) {
+		if err := os.MkdirAll(cacheDir, 0700); err != nil {
+			return nil, fmt.Errorf("failed to create helm cache dir: %w", err)
+		}
+	}
+
+	return &HelmCache{
+		cacheDir: cacheDir,
+	}, nil
+}
+
+// Update adds a new entry to the cache for the current helmRepository.
+func (c *HelmCache) Update(helmRepoNamespace, helmRepoName string, value Data) error {
+	// acquire lock
+	lock := flock.New(filepath.Join(c.cacheDir, lockFilename))
+	defer func() {
+		if err := lock.Unlock(); err != nil {
+			log.Printf("Unable to unlock file %s: %v\n", lockFilename, err)
+		}
+	}()
+
+	// release on timeout
+	ctx, cancel := context.WithTimeout(context.TODO(), time.Second)
+	defer cancel()
+
+	ok, err := lock.TryRLockContext(ctx, 250*time.Millisecond) // try to lock every 1/4 second
+	if !ok {
+		return fmt.Errorf("unable to read lock file %s: %w", lockFilename, err)
+	}
```

Comment:
so if it can't acquire the lock after 1 second, it exits with an error, which will then cause the reconciliation loop to repeat (I believe if the reconciliation loop errors it requeues the resource again, you can configure a delay I believe). Given that repeated reconciliation loops for the same resource can be quite noisy in the logs, and in this case also involves a lot of API calls to fetch the index.yaml, is it worth drastically increasing this timeout? Maybe 1 minute?

Reply:
Yeah, makes sense to increase the timeout.

Reply:
:+1: sorry didn't intend to come across so odd, was just thinking aloud as I typed. Glad it made sense :joy:

Reply:
No worries!! It did. :D I could literally read your mental process while thinking this over and I really appreciate that! :)